pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/PubsubProducer.java

```diff
@@ -0,0 +1,401 @@
+  }
+
+  /**
+   * Tries to close the producer cleanly within the specified timeout. If the close does not complete within the
+   * timeout, fail any pending send requests and force close the producer.
+   */
+  public void close(long timeout, TimeUnit unit) {
+    if (timeout < 0) {
+      throw new IllegalArgumentException("Timeout cannot be negative.");
+    }
+
+    channelUtil.closeChannel();
+    log.debug("Closed producer");
+    closed = true;
+  }
+
+  public static class Builder<K, V> {
+    private final String project;
+    private final Serializer<K> keySerializer;
+    private final Serializer<V> valueSerializer;
+
+    private PubsubChannelUtil channelUtil;
+    private PublisherFutureStub publisher;
+    private int batchSize;
+    private boolean isAcks;
+    private Map<String, List<PubsubMessage>> perTopicBatches;
+    private int maxRequestSize;
+    private Time time;
+
+    public Builder(String project, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+      Preconditions.checkArgument(project != null && keySerializer != null && valueSerializer != null);
+      this.project = project;
+      this.keySerializer = keySerializer;
+      this.valueSerializer = valueSerializer;
+      setDefaults();
+    }
+
+    private void setDefaults() {
+      // this is where to set 'regular' fields w/o side effects
+      this.batchSize = PubsubProducerConfig.DEFAULT_BATCH_SIZE;
+      this.isAcks = PubsubProducerConfig.DEFAULT_ACKS;
+      this.perTopicBatches = Collections.synchronizedMap(new HashMap<>());
+      this.maxRequestSize = PubsubProducerConfig.DEFAULT_MAX_REQUEST_SIZE;
+      this.time = new SystemTime();
+    }
+
+    public Builder publisherFutureStub(PublisherFutureStub val) { publisher = val; return this; }
+
+    public Builder batchSize(int val) {
+      Preconditions.checkArgument(val > 0);
+      batchSize = val;
+      return this;
+    }
+
+    public Builder isAcks(boolean val) { isAcks = val; return this; }
+
+    public Builder perTopicBatches(Map<String, List<PubsubMessage>> val) { perTopicBatches = val; return this; }
+
+    public Builder maxRequestSize(int val) {
+      Preconditions.checkArgument(val >= 0);
+      maxRequestSize = val;
+      return this;
+    }
+
+    public Builder time(Time val) { time = val; return this; }
+
+    public Builder pubsubChannelUtil(PubsubChannelUtil val) { channelUtil = val; return this; }
+
+    public PubsubProducer build() {
+      // this is where to set fields w/ side effects
+      if (channelUtil == null) {
+        this.channelUtil = new PubsubChannelUtil();
+      }
+      if (publisher == null) {
+        this.publisher = PublisherGrpc.newFutureStub(channelUtil.channel()).withCallCredentials(channelUtil.callCredentials());
+      }
+      return new PubsubProducer(this);
+    }
+  }
+
+  /** Taken from KafkaProducer.java since FutureFailure is private inside that class. */
+  private static class FutureFailure implements Future<RecordMetadata> {
```

Comment:
Is there a specific reason we need to use this class as opposed to some other Future class? I'd rather we not pull parts of the private Kafka implementation out verbatim.